<!--
This HTML is auto-generated.  DO NOT EDIT THIS FILE!  If you are writing a new
PEP, see http://www.python.org/dev/peps/pep-0001 for instructions and links
to templates.  DO NOT USE THIS HTML FILE AS YOUR TEMPLATE!
-->
<table class="rfc2822 docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">PEP:</th><td class="field-body">325</td>
</tr>
<tr class="field"><th class="field-name">Title:</th><td class="field-body">Resource-Release Support for Generators</td>
</tr>
<tr class="field"><th class="field-name">Version:</th><td class="field-body">$Revision$</td>
</tr>
<tr class="field"><th class="field-name">Last-Modified:</th><td class="field-body"><a class="reference external" href="https://hg.python.org/peps/file/tip/pep-0325.txt">$Date$</a></td>
</tr>
<tr class="field"><th class="field-name">Author:</th><td class="field-body">Samuele Pedroni &lt;pedronis&#32;&#97;t&#32;python.org&gt;</td>
</tr>
<tr class="field"><th class="field-name">Status:</th><td class="field-body">Rejected</td>
</tr>
<tr class="field"><th class="field-name">Type:</th><td class="field-body">Standards Track</td>
</tr>
<tr class="field"><th class="field-name">Content-Type:</th><td class="field-body"><a class="reference external" href="/dev/peps/pep-0012">text/x-rst</a></td>
</tr>
<tr class="field"><th class="field-name">Created:</th><td class="field-body">25-Aug-2003</td>
</tr>
<tr class="field"><th class="field-name">Python-Version:</th><td class="field-body">2.4</td>
</tr>
<tr class="field"><th class="field-name">Post-History:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
<hr />
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#abstract" id="id5">Abstract</a></li>
<li><a class="reference internal" href="#pronouncement" id="id6">Pronouncement</a></li>
<li><a class="reference internal" href="#rationale" id="id7">Rationale</a></li>
<li><a class="reference internal" href="#possible-semantics" id="id8">Possible Semantics</a></li>
<li><a class="reference internal" href="#remarks" id="id9">Remarks</a></li>
<li><a class="reference internal" href="#open-issues" id="id10">Open Issues</a></li>
<li><a class="reference internal" href="#alternative-ideas" id="id11">Alternative Ideas</a></li>
<li><a class="reference internal" href="#references" id="id12">References</a></li>
<li><a class="reference internal" href="#copyright" id="id13">Copyright</a></li>
</ul>
</div>
<div class="section" id="abstract">
<h1><a class="toc-backref" href="#id5">Abstract</a></h1>
<p>Generators allow for natural coding and abstraction of traversal
over data.  Currently if external resources needing proper timely
release are involved, generators are unfortunately not adequate.
The typical idiom for timely release is not supported, a yield
statement is not allowed in the try clause of a try-finally
statement inside a generator.  The finally clause execution can be
neither guaranteed nor enforced.</p>
<p>This PEP proposes that the built-in generator type implement a
close method and destruction semantics, such that the restriction
on yield placement can be lifted, expanding the applicability of
generators.</p>
</div>
<div class="section" id="pronouncement">
<h1><a class="toc-backref" href="#id6">Pronouncement</a></h1>
<p>Rejected in favor of <a class="reference external" href="/dev/peps/pep-0342">PEP 342</a> which includes substantially all of
the requested behavior in a more refined form.</p>
</div>
<div class="section" id="rationale">
<h1><a class="toc-backref" href="#id7">Rationale</a></h1>
<p>Python generators allow for natural coding of many data traversal
scenarios.  Their instantiation produces iterators,
i.e. first-class objects abstracting traversal (with all the
advantages of first- classness).  In this respect they match in
power and offer some advantages over the approach using iterator
methods taking a (smalltalkish) block.  On the other hand, given
current limitations (no yield allowed in a try clause of a
try-finally inside a generator) the latter approach seems better
suited to encapsulating not only traversal but also exception
handling and proper resource acquisition and release.</p>
<p>Let's consider an example (for simplicity, files in read-mode are
used):</p>
<pre class="literal-block">
def all_lines(index_path):
    for path in file(index_path, &quot;r&quot;):
        for line in file(path.strip(), &quot;r&quot;):
            yield line
</pre>
<p>this is short and to the point, but the try-finally for timely
closing of the files cannot be added.  (While instead of a path, a
file, whose closing then would be responsibility of the caller,
could be passed in as argument, the same is not applicable for the
files opened depending on the contents of the index).</p>
<p>If we want timely release, we have to sacrifice the simplicity and
directness of the generator-only approach: (e.g.)</p>
<pre class="literal-block">
class AllLines:

     def __init__(self, index_path):
         self.index_path = index_path
         self.index = None
         self.document = None

     def __iter__(self):
         self.index = file(self.index_path, &quot;r&quot;)
         for path in self.index:
             self.document = file(path.strip(), &quot;r&quot;)
         for line in self.document:
             yield line
         self.document.close()
         self.document = None

     def close(self):
         if self.index:
             self.index.close()
         if self.document:
             self.document.close()
</pre>
<p>to be used as:</p>
<pre class="literal-block">
all_lines = AllLines(&quot;index.txt&quot;)
try:
    for line in all_lines:
        ...
finally:
    all_lines.close()
</pre>
<p>The more convoluted solution implementing timely release, seems
to offer a precious hint.  What we have done is encapsulate our
traversal in an object (iterator) with a close method.</p>
<p>This PEP proposes that generators should grow such a close method
with such semantics that the example could be rewritten as:</p>
<pre class="literal-block">
# Today this is not valid Python: yield is not allowed between
# try and finally, and generator type instances support no
# close method.

def all_lines(index_path):
    index = file(index_path, &quot;r&quot;)
    try:
        for path in index:
            document = file(path.strip(), &quot;r&quot;)
        try:
            for line in document:
                yield line
        finally:
            document.close()
    finally:
        index.close()

all = all_lines(&quot;index.txt&quot;)
try:
    for line in all:
        ...
finally:
    all.close() # close on generator
</pre>
<p>Currently <a class="reference external" href="/dev/peps/pep-0255">PEP 255</a> <a class="footnote-reference" href="#id3" id="id1">[1]</a> disallows yield inside a try clause of a
try-finally statement, because the execution of the finally clause
cannot be guaranteed as required by try-finally semantics.</p>
<p>The semantics of the proposed close method should be such that
while the finally clause execution still cannot be guaranteed, it
can be enforced when required.  Specifically, the close method
behavior should trigger the execution of the finally clauses
inside the generator, either by forcing a return in the generator
frame or by throwing an exception in it.  In situations requiring
timely resource release, close could then be explicitly invoked.</p>
<p>The semantics of generator destruction on the other hand should be
extended in order to implement a best-effort policy for the
general case.  Specifically, destruction should invoke <tt class="docutils literal">close()</tt>.
The best-effort limitation comes from the fact that the
destructor's execution is not guaranteed in the first place.</p>
<p>This seems to be a reasonable compromise, the resulting global
behavior being similar to that of files and closing.</p>
</div>
<div class="section" id="possible-semantics">
<h1><a class="toc-backref" href="#id8">Possible Semantics</a></h1>
<p>The built-in generator type should have a close method
implemented, which can then be invoked as:</p>
<pre class="literal-block">
gen.close()
</pre>
<p>where <tt class="docutils literal">gen</tt> is an instance of the built-in generator type.
Generator destruction should also invoke close method behavior.</p>
<p>If a generator is already terminated, close should be a no-op.</p>
<p>Otherwise, there are two alternative solutions, Return or
Exception Semantics:</p>
<p>A - Return Semantics: The generator should be resumed, generator
execution should continue as if the instruction at the re-entry
point is a return.  Consequently, finally clauses surrounding the
re-entry point would be executed, in the case of a then allowed
try-yield-finally pattern.</p>
<p>Issues: is it important to be able to distinguish forced
termination by close, normal termination, exception propagation
from generator or generator-called code?  In the normal case it
seems not, finally clauses should be there to work the same in all
these cases, still this semantics could make such a distinction
hard.</p>
<p>Except-clauses, like by a normal return, are not executed, such
clauses in legacy generators expect to be executed for exceptions
raised by the generator or by code called from it.  Not executing
them in the close case seems correct.</p>
<p>B - Exception Semantics: The generator should be resumed and
execution should continue as if a special-purpose exception
(e.g. CloseGenerator) has been raised at re-entry point.  Close
implementation should consume and not propagate further this
exception.</p>
<p>Issues: should <tt class="docutils literal">StopIteration</tt> be reused for this purpose?  Probably
not.  We would like close to be a harmless operation for legacy
generators, which could contain code catching <tt class="docutils literal">StopIteration</tt> to
deal with other generators/iterators.</p>
<p>In general, with exception semantics, it is unclear what to do if
the generator does not terminate or we do not receive the special
exception propagated back.  Other different exceptions should
probably be propagated, but consider this possible legacy
generator code:</p>
<pre class="literal-block">
try:
    ...
    yield ...
    ...
except: # or except Exception:, etc
    raise Exception(&quot;boom&quot;)
</pre>
<p>If close is invoked with the generator suspended after the yield,
the except clause would catch our special purpose exception, so we
would get a different exception propagated back, which in this
case ought to be reasonably consumed and ignored but in general
should be propagated, but separating these scenarios seems hard.</p>
<p>The exception approach has the advantage to let the generator
distinguish between termination cases and have more control.  On
the other hand, clear-cut semantics seem harder to define.</p>
</div>
<div class="section" id="remarks">
<h1><a class="toc-backref" href="#id9">Remarks</a></h1>
<p>If this proposal is accepted, it should become common practice to
document whether a generator acquires resources, so that its close
method ought to be called.  If a generator is no longer used,
calling close should be harmless.</p>
<p>On the other hand, in the typical scenario the code that
instantiated the generator should call close if required by it.
Generic code dealing with iterators/generators instantiated
elsewhere should typically not be littered with close calls.</p>
<p>The rare case of code that has acquired ownership of and need to
properly deal with all of iterators, generators and generators
acquiring resources that need timely release, is easily solved:</p>
<pre class="literal-block">
if hasattr(iterator, 'close'):
    iterator.close()
</pre>
</div>
<div class="section" id="open-issues">
<h1><a class="toc-backref" href="#id10">Open Issues</a></h1>
<p>Definitive semantics ought to be chosen.  Currently Guido favors
Exception Semantics.  If the generator yields a value instead of
terminating, or propagating back the special exception, a special
exception should be raised again on the generator side.</p>
<p>It is still unclear whether spuriously converted special
exceptions (as discussed in Possible Semantics) are a problem and
what to do about them.</p>
<p>Implementation issues should be explored.</p>
</div>
<div class="section" id="alternative-ideas">
<h1><a class="toc-backref" href="#id11">Alternative Ideas</a></h1>
<p>The idea that the yield placement limitation should be removed and
that generator destruction should trigger execution of finally
clauses has been proposed more than once.  Alone it cannot
guarantee that timely release of resources acquired by a generator
can be enforced.</p>
<p><a class="reference external" href="/dev/peps/pep-0288">PEP 288</a> <a class="footnote-reference" href="#id4" id="id2">[2]</a> proposes a more general solution, allowing custom
exception passing to generators.  The proposal in this PEP
addresses more directly the problem of resource release.  Were <a class="reference external" href="/dev/peps/pep-0288">PEP
288</a> implemented, Exceptions Semantics for close could be layered
on top of it, on the other hand <a class="reference external" href="/dev/peps/pep-0288">PEP 288</a> should make a separate
case for the more general functionality.</p>
</div>
<div class="section" id="references">
<h1><a class="toc-backref" href="#id12">References</a></h1>
<table class="docutils footnote" frame="void" id="id3" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[1]</a></td><td><a class="reference external" href="/dev/peps/pep-0255">PEP 255</a> Simple Generators
<a class="reference external" href="http://www.python.org/dev/peps/pep-0255/">http://www.python.org/dev/peps/pep-0255/</a></td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id4" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id2">[2]</a></td><td><a class="reference external" href="/dev/peps/pep-0288">PEP 288</a> Generators Attributes and Exceptions
<a class="reference external" href="http://www.python.org/dev/peps/pep-0288/">http://www.python.org/dev/peps/pep-0288/</a></td></tr>
</tbody>
</table>
</div>
<div class="section" id="copyright">
<h1><a class="toc-backref" href="#id13">Copyright</a></h1>
<p>This document has been placed in the public domain.</p>
<!-- Local Variables:
mode: indented-text
indent-tabs-mode: nil
sentence-end-double-space: t
fill-column: 70
End: -->
</div>

